### ISSUE TITLE
Concurrent task execution

### ISSUE THREAD

**Issue by ixxie** [metaist/ds#92]
Is it possible to execute multiple long-running tasks concurrently, feeding `stdout` and `stderr` into the main process?

Our usecase is running a variety of services in dev mode in tandem, during development. We currently use [concurrently.js](https://github.com/open-cli-tools/concurrently) for this, a NodeJS based tool. The challenge with this tool is that it doesn't play well with Uvicorn's HMR functionality.

I realize this is a big ask, so I wouldn't be surprised if you decide to scope this out of the library, but it _is_ a very important usecase for workspace task running in my opinion.

**Comment by metaist** [comment]
Thanks for writing out your use case; indeed that's a big ask and when I thought about it before, I kept it out of scope because I didn't have a compelling enough need, but I'm willing to think this through.

Syntactically, we'd add a `--parallel` to the CLI and `composite.parallel` to the config. We'd also label lines with the task name that started them + color (or something like that).

But the actual mechanism requires us to know:
- What would we do differently to make sure we don't have the same issues with HMR?
- Is the problem that when a subprocess is restarted the parent process doesn't know about it?
- How should the runner handle errors? Since it's a long-running thing `keep_going` implied? Or maybe if one goes down they all go down (unless you're explicit about which keep running)?

**Comment by ixxie** [comment]
> Thanks for writing out your use case; indeed that's a big ask and when I thought about it before, I kept it out of scope because I didn't have a compelling enough need, but I'm willing to think this through.

Thanks for taking the time!

>  What would we do differently to make sure we don't have the same issues with HMR?

Unfortunately, my knowledge of processes is limited. 

>  Is the problem that when a subprocess is restarted the parent process doesn't know about it?

At least `concurrently` seems to have a way to know about it; it offers the following options:

> `restartTries`: how many attempts to restart a process that dies will be made. Default: 0.
> `restartDelay`: how many milliseconds to wait between process restarts. Default: 0.

>  How should the runner handle errors? Since it's a long-running thing `keep_going` implied? Or maybe if one goes down they all go down (unless you're explicit about which keep running)?

In concurrently, there are the following options:

> `killOthers`: an array of exitting conditions that will cause a process to kill others. Can contain any of success or failure.
> `successCondition`: the condition to consider the run was successful. If first, only the first process to exit will make up the success of the run; if last, the last process that exits will determine whether the run succeeds. Anything else means all processes should exit successfully.

**Comment by metaist** [comment]
In your use case what values do you assign to these configuration options? And what other options are relevant? How many tasks do you typically run?

I have an idea for a basic parallel runner (run each task in a `subprocess` and read all the `stdout` and `stderr` in separate threads so the lines don't overlap), but it's not likely to be super configurable.

Given how the rest of `ds` works, here are my thoughts on `concurrently`  options:
- `cwd`: ds already supports this, it should work the same
- `defaultInputTarget`: I would not be reading from proc.stdin, so there would be no input target. I would just read from `process.stdout/.stdout` and write them to the main process `stdout/stderr`; you can redirect that in the command in the usual bash ways.
- `handleInput`: false; I don't get this use case.
- `inputStream`: null; I wouldn't read from stdin
- `pauseInputStreamOnFinish`: false; not reading from stdin
- `killOthers`: if a process depends on some other process, we can use the regular chain/composite syntax. We would just need a way of saying "runs in the background". If it returns a non-zero error code, that's the signal to the parent to stop running (like breaking the normal chain). Unless `keep_going=true` which, like usual, doesn't stop any other process.
- `maxProcesses`: one per task.
- `outputStream`: I would just write to the main `stdout`.
- `prefix`: I would just use the task name or an incrementing number for anonymous tasks.
- `prefixColors`: I would use a fixed list with reasonable contrast and loop over them when I ran out.
- `prefixLength`: I would either have a fixed length (so output lines are aligned) or "as long as the name".
- `raw`: I probably wouldn't offer this option (maybe  `--no-color` because that's pretty standard)
- `successCondition`: if keep_going always successful (like in non-parallel version). Otherwise, probably all need to return successfully. 
- `restartTries`: 0. Restarting a dead process feels like a very different kind of tool, but I get that this is the specific thing you're asking for.
- `restartDelay`: n/a. But zero or 0.1s if I did implement it.
- `timestampFormat`: n/a, I'd prefix with name or number.

**Comment by ixxie** [comment]
We run about 2-5 tasks in tandem, depending on what we are developing. Think small-to-medium complexity web app development: we want to do something like what docker compose does, but without the cruft of containers. We typically always run the Uvicorn backend, and a Vite frontend dev server, and sometimes also 1-3 other Python services.

We barely use any of Concurrently's options. We use `--raw` because otherwise color output is broken. I experimented with `--restart-tries` and `--kill-others` in an attempt to get HMR going, but had no luck. 

I do think some API for controlling what happens if one process dies is probably important, but it seems the API you propose seems quite flexible and should address most usecases. For now, in our usecase, I think it suffices to have the main process exit if any process dies.

We did use prefixes for a while, but in the end we wanted more control and just decided to print identifiers ourselves. My take is that its better to have _no_ prefixes than _unconfigurable_ prefixes, since that allows users to do what we did without adding extra clutter. Of course, _optional_ prefixes of any kind never hurt.

I wonder if its sometimes beneficial to have the option to redirect (one or all) task's `stderr` to the main process `stdout` instead of its `stderr`, but I'm not sure about this. 

Another thing I forgot to mention which Concurrently does quite well, is handling of keyboard interrupt. This is important.

_tl;dr_ - the important stuff for our usecase:
 - Basic control over how task exits propagate
 - Keyboard interrupt stops everything
 - HMR for Vite and Uvicorn "just works"
 - Redirect stdout/stderr to parent process
 - Support modern color terminal output
 - Allow having no prefixes on the output

**Comment by metaist** [comment]
This all makes sense. I wrote a minimal test for uvicorn and I can get it to work in parallel with another task.

Next week, I'll try a few more tests and then add an experimental flag.

**Comment by metaist** [comment]
Ok, I have an experimental top-level `--parallel` flag basically working. I learned a bunch about how the `uvicorn` reloading (and `vite` reloading) work. Basically they start a process that starts its own children. When they detect a file change, they kill the relevant child process and restart it. So it doesn't really impact the parent (which we interact with).

I tested out a `uvicorn` + `vite` setup and it's quite nice running both the front-end and server from a single command and having them both reload as needed. 

For your use case:
- Basic control over how task exits propagate: For this initial experiment, I just ignore all errors. No process takes down others. (I'm thinking about how I'd need to change parsing / running to handle this properly.)
- Keyboard interrupt stops everything: Yes.
- HMR for Vite and Uvicorn "just works": Yes.
- Redirect stdout/stderr to parent process: Yes.
- Support modern color terminal output: Yes (at least Vite and Uvicorn display colors correctly)
- Allow having no prefixes on the output: easiest way was to have no prefixes

As simple test case:

```bash
ds --parallel 'ping -c 4 google.com' 'ping -c 4 example.com'
```

For my `uvicorn + vite` test, I ran:
```bash
ds --parallel --no-config ./server.py vite
```
```text
$ ds --no-config --parallel ./server.py vite

> ds ./server.py
$ ./server.py
WARNING: EXPERIMENTAL: running tasks in parallel

> ds vite
$ vite

  VITE v5.4.7  ready in 275 ms

  ➜  Local:   http://localhost:5173/
  ➜  Network: use --host to expose
  ➜  press h + enter to show help
starting server
INFO:     Will watch for changes in these directories: ['~/Documents/code/tmp']
INFO:     Uvicorn running on http://127.0.0.1:8080 (Press CTRL+C to quit)
INFO:     Started reloader process [130343] using StatReload
INFO:     Started server process [130373]
INFO:     Waiting for application startup.
INFO:     Application startup complete.
INFO:     127.0.0.1:47668 - "GET / HTTP/1.1" 200 OK
INFO:     127.0.0.1:48170 - "GET / HTTP/1.1" 200 OK
WARNING:  StatReload detected changes in 'server.py'. Reloading...
INFO:     Shutting down
INFO:     Waiting for application shutdown.
INFO:     Application shutdown complete.
INFO:     Finished server process [130373]
INFO:     Started server process [130603]
INFO:     Waiting for application startup.
INFO:     Application startup complete.
INFO:     127.0.0.1:60620 - "GET / HTTP/1.1" 200 OK
```

**Comment by ixxie** [comment]
Wow, amazing @metaist! Thank you! I'm grateful you took this on so fast. I'm going to experiment with this in the coming days and see how it plays in our system.

**Comment by ixxie** [comment]
@metaist I got the command going pretty much as expected, except the HMR is still not working. What I realized (thanks to your work) is that the issue is on our end, not due to `concurrently` or `ds`. We have a custom CLI for managing our dev environment, which:

- Injects some configuration as environment variables
- Uses `subprocess.run` to run ds or concurrently, which in turn run the tasks
- Runs `poetry` or `npm` script, excuted as a shell command (e.g. `cd backend && poetry run api`)

Whenever HMR is triggered we get a message `Terminate batch job (Y/N)?`. From [this stackexchange question](https://superuser.com/questions/35698/how-to-supress-terminate-batch-job-y-n-confirmation) I gather that the origin of this is actually `cmd.exe`. But I'm not sure how to prevent this from happening.

**Comment by ixxie** [comment]
Digging further, I found [a discussion of this issue in Uvicorn's repo](https://github.com/encode/uvicorn/discussions/2292).

I'm sorry if I've wasted your time here!

**Comment by metaist** [comment]
No worries. Thanks for digging and getting back to me.

I'm still going to move forward with `--parallel` because the syntax is simpler than GNU Parallel and I'm already finding it useful.

Below was written before your latest message (I think `ds` can basically do what your custom CLI does).

---

Ah, Windows. This is a little harder for me to debug, but `ds` might be able to help here too (by replacing your custom runner).

Assuming you have a file layout like:
```text
project
 \backend
  \.venv
  \.env
  \pyproject.toml (with something like `tool.poetry.scripts.api = "api:run"`)
 \frontend
  \node_modules
  \package.json
```

Update `backend/pyproject.toml` to add:
```toml
[tool.ds.scripts]
run.shell = "poetry run api" # ds activates the venv, so this could be just "api"
run.env-file = ".env"
```

Update `frontend/package.json` to add:
```json
{ 
  "scripts": {
   "run": "vite"
  }
}
```

Now add a `project/ds.toml` to bring the workspace together:

```toml
[workspace]
members = ["backend", "frontend"]
```

Now you could run `ds -w* run` which will `cd` into each directory and run the `run` command there.

**Comment by metaist** [comment]
By the way, have you tried running your stack using `PowerShell`? From what I've seen, it handles things a bit differently that `cmd.exe` (usually for the better).

**Comment by ixxie** [comment]
> By the way, have you tried running your stack using `PowerShell`? From what I've seen, it handles things a bit differently that `cmd.exe` (usually for the better).

Yeah we use PowerShell 7.

> Ah, Windows. This is a little harder for me to debug, but ds might be able to help here too (by replacing your custom runner).

All together replacing our CLI is out of the question, it does a lot of stuff `ds` can't do (we have a custom config `toml` of our own, and a bunch of other complex stuff handled by the CLI).

We would be very interested in using `ds` has the task runner, but it seems what I fail to figure out is how to call `ds` via `subprocess` such that it doesn't run into the [cursed Windows process killing issue](https://discuss.python.org/t/terminateprocess-via-os-kill-on-windows/30882/5). But it _should_ be possible because effectively, you have done it _in_ the codebase of `ds`.

**Comment by metaist** [comment]
Thanks for trying out the experiments. I see your comments in the `uvicorn` discussion and appreciate how frustrating this is.

**Comment by ixxie** [comment]
> Thanks for trying out the experiments. I see your comments in the `uvicorn` discussion and appreciate how frustrating this is.

Thank you. I guess the last-ditch effort would be to get this running in WSL or something. Luckily, I'm finally buying myself a new Linux box and I won't have to deal with Windows myself for much longer.

**Comment by metaist** [comment]
## Code Review Notes (via Claude)

Before adding config-file support for `parallel`, I reviewed the current implementation in `runner.py` and found several concerns:

### Issue 1: `parallel` propagates to all descendants

In `run()` (lines 170, 178):
```python
resolved = replace(override, ..., parallel=override.parallel or task.parallel)
...
for dep in task.depends:
    self.run(dep, resolved)  # passes parallel=true to ALL children
```

If task A has `parallel=true` and depends on B which depends on [D, E], then D and E will **also** run in parallel, even if B should be sequential.

**Expected behavior:** `parallel` should only affect direct children, not be inherited by grandchildren.

### Issue 2: No sync point after parallel children

After spawning parallel dependencies, execution immediately continues. For CLI `--parallel` this works because `main()` waits at the end:
```python
for proc in runner.processes:
    proc.wait()
```

But for config-level parallel:
```toml
build = { composite = ["compile_a", "compile_b"], parallel = true }
test = "pytest"  
all = ["build", "test"]
```

Running `ds all` would start `compile_a` and `compile_b`, then **immediately** run `test` without waiting for builds to complete.

**Expected behavior:** Parent task should wait for its parallel children to complete before the parent itself completes.

### Issue 3: No error handling for parallel tasks

Lines 237-252 spawn `Popen` but never check return codes. Parallel tasks effectively have implicit `keep_going=true`. Failures are silently ignored.

**Expected behavior:** Need to decide on error semantics - fail-fast, collect errors, or explicit `keep_going` required?

### Suggested fixes

1. **Don't propagate `parallel` to children** - Remove `parallel` from the `override` passed to dependencies
2. **Add sync point** - After spawning parallel dependencies, wait for them before the parent continues
3. **Error collection** - Track parallel task failures and report/handle them appropriately

**Comment by metaist** [comment]
## Implemented (via Claude)

The `--parallel` flag now works correctly and can also be set in config files:

```toml
[scripts]
dev = { composite = ["backend", "frontend"], parallel = true }
```

### Fixes included:

1. **Parallel propagation** - `parallel` only affects direct children, not grandchildren
2. **Sync point** - Parent waits for parallel children to complete before continuing
3. **Error handling** - Failures are reported with command info, `keep_going` is respected per-task
4. **Config parsing** - `parallel = true` can be set in config files

### Commits:
- 96353bb fix: parallel flag no longer propagates to grandchildren
- 3047213 fix: add sync point for parallel task children
- 2487e3e fix: add error handling for parallel tasks
- b0d38fd add: parse parallel option from config files
